agent/agent.go

```diff
@@ -0,0 +1,287 @@
+		"privilege": a.privilege,
+		"exe_name": a.exe_name,
+	}
+}
+
+// Return minimal subset of agent profile.
+func (a *Agent) GetTrimmedProfile() map[string]interface{} {
+	return map[string]interface{}{
+		"paw": a.paw,
+		"server": a.server,
+		"platform": a.platform,
+		"host": a.host,
+	}
+}
+
+// Pings C2 for instructions and returns them.
+func (a *Agent) Beacon() map[string]interface{} {
+	var beacon map[string]interface{}
+	profile := a.GetFullProfile()
+	response := a.beaconContact.GetBeaconBytes(profile)
+	if response != nil {
+		beacon = processBeacon(response)
+	} else {
+		output.VerbosePrint("[-] beacon: DEAD")
+	}
+	return beacon
+}
+
+// Converts the given data into a beacon with instructions.
+func processBeacon(data []byte) map[string]interface{} {
+	var beacon map[string]interface{}
+	if err := json.Unmarshal(data, &beacon); err != nil {
+		output.VerbosePrint(fmt.Sprintf("[-] Malformed beacon received: %s", err.Error()))
+	} else {
+		var commands interface{}
+		if err := json.Unmarshal([]byte(beacon["instructions"].(string)), &commands); err != nil {
+			output.VerbosePrint(fmt.Sprintf("[-] Malformed beacon instructions received: %s", err.Error()))
+		} else {
+			output.VerbosePrint("[+] beacon: ALIVE")
+			beacon["sleep"] = int(beacon["sleep"].(float64))
+			beacon["watchdog"] = int(beacon["watchdog"].(float64))
+			beacon["instructions"] = commands
+		}
+	}
+	return beacon
+}
+
+func (a *Agent) Heartbeat() {
+	// Add any heartbeat functionality here.
+}
+
+func (a *Agent) Terminate() {
+	// Add any cleanup/termination functionality here.
+	output.VerbosePrint("[*] Terminating Sandcat Agent... goodbye.")
+}
+
+// Runs a single instruction and send results.
+func (a *Agent) RunInstruction(command map[string]interface{}, payloads []string) {
+	timeout := int(command["timeout"].(float64))
+	result := make(map[string]interface{})
+	commandOutput, status, pid := execute.RunCommand(command["command"].(string), payloads, command["executor"].(string), timeout)
+	result["id"] = command["id"]
+	result["output"] = commandOutput
+	result["status"] = status
+	result["pid"] = pid
+ 	a.beaconContact.SendExecutionResults(a.GetTrimmedProfile(), result)
+}
+
+// Sets the C2 communication channels for the agent according to the specified C2 configuration map.
+// Will default to HTTP if requested C2 is not available or its requirements aren't met. If defaulting
+// to HTTP when it is not available,or if no communication channels are available at all, an error will be returned.
+func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
+	if len(contact.CommunicationChannels) > 0 {
+		if requestedC2, ok := c2Config["c2Name"]; ok {
+			if err := a.attemptSelectChannel(c2Config, requestedC2); err == nil {
+				return nil
+			}
+		}
+		output.VerbosePrint(fmt.Sprintf("[*] Attempting to fall back to default C2 channel %s", a.defaultC2))
+		if err := a.attemptSelectChannel(c2Config, a.defaultC2); err == nil {
+			return nil
+		}
```

Comment:
my gut tells me the solution I came up with is wrong because we are reusing code, lets refactor this somehow... maybe a separate functions or a better way to flow the code overall

Reply:
How's this?

```
func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
	if len(contact.CommunicationChannels) > 0 {
		if requestedC2, ok := c2Config["c2Name"]; ok {
			if err := a.attemptSelectChannel(c2Config, requestedC2); err == nil {
				return nil
			} else {
				output.VerbosePrint(fmt.Sprintf("[-] %s", err.Error()))
			}
		} else {
			output.VerbosePrint("[-] No requested C2 channel provided.")
		}
		output.VerbosePrint(fmt.Sprintf("[*] Attempting to fall back to default C2 channel %s", a.defaultC2))
		return a.attemptSelectChannel(c2Config, a.defaultC2)
	}
	return errors.New("No possible C2 communication channels found.")
}
```